src/common/dng_opcode.c

```diff
@@ -0,0 +1,115 @@
+/*
+    This file is part of darktable,
+    Copyright (C) 2011-2021 darktable developers.
+
+    darktable is free software: you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation, either version 3 of the License, or
+    (at your option) any later version.
+
+    darktable is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with darktable.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <glib.h>
+#include <stdint.h>
+#include <stdio.h>
+
+#include "dng_opcode.h"
+#include "image.h"
+
+#define OPCODE_ID_GAINMAP (9)
+
+static double get_double(uint8_t *ptr)
+{
+  double v;
+  ((uint8_t *)&v)[0] = ptr[7];
+  ((uint8_t *)&v)[1] = ptr[6];
+  ((uint8_t *)&v)[2] = ptr[5];
+  ((uint8_t *)&v)[3] = ptr[4];
+  ((uint8_t *)&v)[4] = ptr[3];
+  ((uint8_t *)&v)[5] = ptr[2];
+  ((uint8_t *)&v)[6] = ptr[1];
+  ((uint8_t *)&v)[7] = ptr[0];
+  return v;
+}
+
+static float get_float(uint8_t *ptr)
+{
+  float v;
+  ((uint8_t *)&v)[0] = ptr[3];
+  ((uint8_t *)&v)[1] = ptr[2];
+  ((uint8_t *)&v)[2] = ptr[1];
+  ((uint8_t *)&v)[3] = ptr[0];
+  return v;
+}
+
+static uint32_t get_long(uint8_t *ptr)
+{
+  uint32_t v;
+  ((uint8_t *)&v)[0] = ptr[3];
+  ((uint8_t *)&v)[1] = ptr[2];
+  ((uint8_t *)&v)[2] = ptr[1];
+  ((uint8_t *)&v)[3] = ptr[0];
+  return v;
+}
```

Comment:
Ha, GLib actually provides all of this already, see `g_ntohl` (alias of `GUINT32_FROM_BE()`) and `GUINT64_FROM_BE()`

Reply:
I didn't try to support big endian, since the entire application only builds on little endian systems: https://github.com/darktable-org/darktable/blob/83a96dfda12ea22f3ba2c0e0333174964294c98d/ConfigureChecks.cmake#L81

Reply:
Nonetheless, I'd prefer use of existing functions here that have a known implementation via builtins, one less thing to worry about.

Reply:
Just pushed a fix to use the glib macros.

Reply:
Nice, thanks!